src/stream/src/executor/dynamic_filter.rs

```diff
@@ -332,12 +351,23 @@ impl<S: StateStore, const USE_WATERMARK_CACHE: bool> DynamicFilterExecutor<S, US
                     let chunk = chunk.compact(); // Is this unnecessary work?
 
                     // The condition is `None` if it is always false by virtue of a NULL right
-                    // input, so we save evaluating it on the datachunk
+                    // input, so we save evaluating it on the datachunk.
                     let filter_condition =
-                        build_cond(committed_rhs_value.clone().flatten()).transpose()?;
-
-                    let (new_ops, new_visibility) =
-                        self.apply_batch(&chunk, filter_condition).await?;
+                        build_cond(self.comparator, committed_rhs_value.clone().flatten())
+                            .transpose()?;
+
+                    // The condition is `None` if there's no committed state cleaning watermark before.
+                    // Note that we should not use `state_cleaning_watermark` variable here, because
+                    // it represents outstanding watermark to be applied. Here we need the watermark
+                    // that has been applied to the state table, just like why we use `prev_epoch_value`
+                    // instead of `current_epoch_value` for `filter_condition`.
+                    let below_watermark_condition =
+                        build_cond(LessThan, self.left_table.get_committed_watermark().cloned())
```

Comment:
Thanks to the removal of "state clean watermark buffer", now we can get up-to-date committed watermark from state table (and even after recovery).